Microsoft.Toolkit.HighPerformance/Memory/Memory2D{T}.cs

```diff
@@ -0,0 +1,238 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Buffers;
+using System.ComponentModel;
+using System.Diagnostics.Contracts;
+using System.Runtime.CompilerServices;
+using System.Runtime.InteropServices;
+using Microsoft.Toolkit.HighPerformance.Extensions;
+
+namespace Microsoft.Toolkit.HighPerformance.Memory
+{
+    /// <summary>
+    /// <see cref="Memory2D{T}"/> represents a 2D region of arbitrary memory. It is to <see cref="Span2D{T}"/>
+    /// what <see cref="Memory{T}"/> is to <see cref="Span{T}"/>. For further details on how the internal layout
+    /// is structured, see the docs for <see cref="Span2D{T}"/>. The <see cref="Memory2D{T}"/> type can wrap arrays
+    /// of any rank, provided that a valid series of parameters for the target memory area(s) are provided.
+    /// </summary>
+    /// <typeparam name="T">The type of items in the current <see cref="Span2D{T}"/> instance.</typeparam>
+    public readonly struct Memory2D<T> : IEquatable<Memory2D<T>>
+    {
+        /// <summary>
+        /// The target <see cref="Array"/> instance, if present.
+        /// </summary>
+        private readonly Array? array;
```

Comment:
`object` should be used, even if `Array` is currently the only supported type, to keep parity with `Memory` and allow future support for `MemoryManager<T>` or a `MemoryManager2D<T>` type

Reply:
Done in a past commit, and also added support for more wrapped types 👍